tutorials/03_moduly.md

```diff
@@ -0,0 +1,290 @@
+
+setup.py
+--------
+
+Základním stavebním kamenem Python balíčku je soubor `setup.py`, který
+obsahuje všechna potřebná metadata ve volání funkce `setup()` z modulu
+`setuptools`.
+
+Pojďme vytvořit jeho minimální variantu:
+
+```python
+from setuptools import setup
+
+
+setup(
+    name='isholiday',
+    version='0.1',
+    description='Finds Czech holiday for given year',
+    author='Ondřej Caletka',
+    author_email='ondrej@caletka.cz',
+    license='Public Domain',
+    url='https://gist.github.com/oskar456/e91ef3ff77476b0dbc4ac19875d0555e',
+    py_modules=['isholiday'],
+)
+```
+
+Všimněte si, že jsme balíček pojmenoval stejně jako soubor se zdrojovým kódem.
+Je to dobrá konvence, ale není to technicky nutné.
+
+Balíček můžeme zkusit nainstalovat do virtualenvu:
+
+```bash
+$ python3.5 -m venv env
+$ . env/bin/activate
+(env)$ python setup.py install
+...
+(env)$ python
+>>> import isholiday
+>>> 
+(env)$ python -m pip freeze
+isholiday==0.1
+```
+
+Přes `setup.py` můžeme dělat další věci, například vytvořit archiv s balíčkem:
+
+```bash
+(env)$ python3 setup.py sdist
+...
+warning: sdist: standard file not found: should have one of README, README.rst, README.txt
+...
+```
+
+Extra soubory do zdrojového balíčku
+-----------------------------------
+
+Jak vidíte, `setuptools` si stěžuje, že náš projekt nemá `README`.
+Můžeme jej vytvořit a uložit jako `README` přímo v kořenovém adresáři projektu,
+tedy tam, kde byste jej nejspíš čekali.
+
+```
+Czech public holiday checker...
+```
+
+```bash
+(env)$ python3 setup.py sdist
+```
+
+V adresáři `dist` najdete archiv, jeho obsah můžete zkontrolovat, měl by tam
+být i soubor `README`.
+
+Skvělé, pojďme vytvořit i další speciální soubor, `LICENSE`, který bude
+obsahovat text licence, v tomto případě Public Domain.
+Obsah najdete třeba na [CC0].
+
+[CC0]: https://creativecommons.org/publicdomain/zero/1.0/legalcode.txt
+
+Pokud ale se souborem `LICENSE` vytvoříte zdrojový balíček, soubor v archivu
+nebude. Je to proto, že se standardně do archivu přidávají jen některé soubory.
+Další soubory lze přidat pomocí souboru `MANIFEST.in`, dle [dokumentace].
+
+[dokumentace]: https://docs.python.org/3/distutils/sourcedist.html#specifying-the-files-to-distribute
+
+V tomto případě takto:
+
+```
+include LICENSE
+```
+```bash
+(env)$ python3 setup.py sdist
+...
+hard linking LICENSE -> isholiday-0.1
+hard linking MANIFEST.in -> isholiday-0.1
+hard linking README -> isholiday-0.1
+...
+```
+
+Více argumentů pro setup()
+--------------------------
+
+Na chvíli se vrátíme k volání funkce `setup()` a přidáme co nejvíc dalších
+položek ([jejich vysvětlení](https://packaging.python.org/distributing/#setup-args)).
+
+```python
+from setuptools import setup
+
+
+setup(
+    name='isholiday',
+    version='0.1',
+    description='Finds Czech holiday for given year',
+    long_description=''.join(open('README').readlines()),
```

Comment:
Staré https://pypi.python.org/ podporuje jen RST bordel. Markdown je možné automaticky konvertovat do RST pomocí pandocu, viz [moje odpověď na Have the same README both in Markdown and reStructuredText](http://stackoverflow.com/questions/10718767/have-the-same-readme-both-in-markdown-and-restructuredtext/23265673#23265673).